cpu/esp32/periph/flash.c

```diff
@@ -182,9 +178,17 @@ esp_err_t IRAM_ATTR spi_flash_read(size_t addr, void *buff, size_t size)
         uint32_t len_in_word = 4 - pos_in_word;
         len_in_word = (len_in_word < len) ? len_in_word : len;
 
+        /* disable interrupts and the cache */
+        critical_enter();
+        Cache_Read_Disable(PRO_CPU_NUM);
```

Comment:
Instead of copy & pasting that everywhere, just make a `static inline` function for that. `_flash_begin()`/`_flash_end()` maybe

Reply:
This will not work since the macro `critical_enter` declares a local `state` variable that is used by `critical_exit` to restore the interrupt level.
```
#define critical_enter()   int _irq_state = irq_disable()
#define critical_exit()    irq_restore(_irq_state)
```

Reply:
So you can't enter critical sections more than once in a function :\

I don't like macro use when it's avoidable, but this is unrelated to this PR.

You could still double down on the macro use and `#define flash_begin() critical_enter(); Cache_Read_Disable(PRO_CPU_NUM)`

Reply:
> So you can't enter critical sections more than once in a scope :\

Since an interrupt-based context switch is used, it can not happen that the same area is entered a second time. Simply disabling interrupts is a very basic critical section mechanism (which is also used by default in FreeRTOS), but using mutexes wouldn't work in the interrupt context.

> I don't like macro use when it's avoidable, but this is unrelated to this PR.

These macros are a bit historical. In a former implementation, it made a bit more and vendor files required the their definition. Since it is not possible to realize them as static inline function, they were declared as macros.

> You could still double down on the macro use and `#define flash_begin() critical_enter(); Cache_Read_Disable(PRO_CPU_NUM)`

I would prefer to leave them separate to make it more clear, what's done.